src/v/cluster/data_migration_backend.cc

```diff
@@ -1251,20 +1250,21 @@ ss::future<> backend::reconcile_migration(
       metadata.id,
       mrstate.scope.sought_state);
     co_await std::visit(
-      [this, &metadata, &mrstate](const auto& migration) mutable {
+      [this, migration_id = metadata.id, &mrstate](
+        const auto& migration) mutable {
           return ss::do_with(
-            migration.topic_nts(),
-            [this, &metadata, &mrstate](const auto& nts) {
-                // poor man's `nts | std::views::enumerate`
-                auto enumerated_nts = std::views::transform(
-                  nts, [index = -1](const auto& nt) mutable {
-                      return std::forward_as_tuple(++index, nt);
-                  });
-                return ssx::async_for_each(
+            // poor man's `migration.topic_nts() | std::views::enumerate`
+            std::views::transform(
+              migration.topic_nts(),
+              [index = -1](const auto& nt) mutable {
+                  return std::forward_as_tuple(++index, nt);
+              }),
+            [this, migration_id, &mrstate](auto& enumerated_nts) {
+                return ss::do_for_each(
                   enumerated_nts,
-                  [this, &metadata, &mrstate](const auto& idx_nt) {
+                  [this, migration_id, &mrstate](const auto& idx_nt) {
                       auto& [idx, nt] = idx_nt;
-                      return reconcile_topic(metadata.id, idx, nt, mrstate);
+                      return reconcile_topic(migration_id, idx, nt, mrstate);
```

Comment:
maybe split out the reformatting? it's kinda hard to tell what is changing here

Reply:
I tried playing with formatting, but it did not make things any cleaner. I'd suggest switching to side-by-side mode for this change.

Essentially, since `ss::do_for_each` does not extend the container's lifetime, we should pass exactly its container argument via `ss::do_with`, that's what the change is about. I can create `ssx::do_for_each_with` if you think it has potential for reuse.